### PR TITLE
[SPARK-43779][SQL] ParseToDate should load the EvalMode in main thread

### PR DESCRIPTION
**Pull request by amaliujia**
### What changes were proposed in this pull request?

ParseToDate should load the EvalMode in main thread instead of loading it in a lazy val. 

### Why are the changes needed?

This is because it is sometimes hard to estimate when the lazy val is executed while the SQLConf where we load the EvalMode is thread local. 

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Existing UT

### CODE REVIEW

**Comment by amaliujia** [comment]
I needed to re-generate the golden files.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala [hunk 1]
```diff
@@ -2044,12 +2044,13 @@ case class MonthsBetween(
 case class ParseToDate(
     left: Expression,
     format: Option[Expression],
-    timeZoneId: Option[String] = None)
+    timeZoneId: Option[String] = None,
+    evalMode: EvalMode.Value = EvalMode.fromSQLConf(SQLConf.get))
   extends RuntimeReplaceable with ImplicitCastInputTypes with TimeZoneAwareExpression {
 
   override lazy val replacement: Expression = format.map { f =>
     Cast(GetTimestamp(left, f, TimestampType, timeZoneId), DateType, timeZoneId)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala:2052]: can we pass evalMode to cast here as well?

File: sql/core/src/test/resources/sql-tests/results/subquery/exists-subquery/exists-within-and-or.sql.out [hunk 2]
```diff
@@ -119,16 +119,10 @@ WHERE  ( NOT EXISTS (SELECT *
...
 -- !query output
-emp 1	10.0
-emp 1	20.0
-emp 2	100.0
-emp 2	300.0
-emp 3	300.0
-emp 4	100.0
-emp 5	1000.0
-emp 6 - no dept	500.0
+java.lang.IllegalArgumentException
```
[the remark this answers is not in the document]
Comment (amaliujia) [reply sql/core/src/test/resources/sql-tests/results/subquery/exists-subquery/exists-within-and-or.sql.out:124]: Auto golden file regeneration gives this update

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala (hunk 1 above)
Comment (amaliujia) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala]: done

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala [hunk 3]
```diff
@@ -2044,12 +2044,14 @@ case class MonthsBetween(
...
     format: Option[Expression],
-    timeZoneId: Option[String] = None)
+    timeZoneId: Option[String] = None,
+    ansiEnabled: Boolean = SQLConf.get.ansiEnabled)
   extends RuntimeReplaceable with ImplicitCastInputTypes with TimeZoneAwareExpression {
 
   override lazy val replacement: Expression = format.map { f =>
-    Cast(GetTimestamp(left, f, TimestampType, timeZoneId), DateType, timeZoneId)
-  }.getOrElse(Cast(left, DateType, timeZoneId)) // backwards compatibility
+    Cast(GetTimestamp(left, f, TimestampType, timeZoneId, ansiEnabled), DateType, timeZoneId)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala:2052]: the Cast here still doesn't have the evalMod passed... But it's a good catch that we should pass ansiEnabled to GetTimestamp

Comment (amaliujia) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala:2052]: hmmm oops I see what is missing. Added.

**Review by cloud-fan — APPROVED** [review]

**Review by cloud-fan — APPROVED** [review]

**Comment by cloud-fan** [comment]
thanks, merging to master!

**Pull request closed without merging**